lib/mongo/server/connection_pool.rb

```diff
@@ -73,19 +119,79 @@ def checkin(connection)
       #
       # @since 2.0.0
       def checkout
-        queue.dequeue
+        raise_if_closed!
+
+        publish_cmap_event(
+          Monitoring::Event::ConnectionCheckoutStarted.new(address)
+        )
+
+        deadline = Time.now + wait_timeout
+        @wait_queue.enter_wait_queue(wait_timeout, deadline) do
+          connections.pop(deadline).tap do |c|
+            publish_cmap_event(
+              Monitoring::Event::ConnectionCheckedOut.new(address, c.id),
+            )
+          end
+        end
+      rescue Error::WaitQueueTimeout
+        publish_cmap_event(
+            Monitoring::Event::ConnectionCheckoutFailed.new(
+              Monitoring::Event::ConnectionCheckoutFailed::TIMEOUT,
+              address,
+            ),
+        )
+        raise
       end
 
       # Disconnect the connection pool.
       #
-      # @example Disconnect the connection pool.
+      # @example Disconnect connection pool.
       #   pool.disconnect!
       #
       # @return [ true ] true.
       #
       # @since 2.1.0
       def disconnect!
-        queue.disconnect!
+        unless closed?
+          connections.disconnect!
+
+          publish_cmap_event(
+            Monitoring::Event::PoolClosed.new(address)
+          )
+        end
+
+        true
+      end
+
+      # Updates the generation number. The connections will be disconnected and removed lazily
+      # when the queue attempts to dequeue them.
+      #
+      # @return [ true ] true.
+      #
+      # @since 2.7.0
+      def clear!
+        unless closed?
+          connections.clear!
+
+          publish_cmap_event(
+            Monitoring::Event::PoolCleared.new(address)
+          )
+        end
+
+        true
+      end
+
+      # Disconnects the pool and prevents any more connections from being checked out afterwards.
+      # If #checkout is called after #close!, a Mongo::Error::PoolClosed error will be raised.
+      #
+      # @return [ true ] true.
+      #
+      # @since 2.7.0
+      def close!
```

Comment:
I believe the first thing this method needs to do is to set the close flag, otherwise concurrent operations in other threads may put something in the available stack for example while the pool is being closed.

Reply:
Sounds good; I'll update it. The reason I had done it this way was that I was using `@connections == nil` as the way to check if the pool was closed, and `@connections` is needed for the call to `disconnect!`.